File: pubnub-java/src/test/java/com/pubnub/api/endpoints/remoteaction/ComposableRemoteActionTest.java
package com.pubnub.api.endpoints.remoteaction;

import com.pubnub.api.PubNubException;
import com.pubnub.api.callbacks.PNCallback;
import com.pubnub.api.models.consumer.PNStatus;
import org.jetbrains.annotations.NotNull;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import static com.pubnub.api.endpoints.remoteaction.ComposableRemoteAction.firstDo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class ComposableRemoteActionTest {

    @Test
    public void sync_happyPath() throws PubNubException {
        //given
        RemoteAction<Integer> composedAction = firstDo(TestRemoteAction.successful(668))
                .then(integerResult -> TestRemoteAction.successful(integerResult * 2))
                .then(integerResult -> TestRemoteAction.successful(integerResult + 1));

        //when
        int result = composedAction.sync();

        //then
        assertEquals(1337, result);
    }


    @Test
    public void async_happyPath() throws InterruptedException {
        //given
        final CountDownLatch latch = new CountDownLatch(1);
        final AtomicInteger result = new AtomicInteger(0);
        RemoteAction<Integer> composedAction = firstDo(TestRemoteAction.successful(668))
                .then(integerResult -> TestRemoteAction.successful(integerResult * 2))
                .then(integerResult -> TestRemoteAction.successful(integerResult + 1));

        //when
        composedAction.async((r, s) -> {
            if (r != null) {
                result.set(r);
                latch.countDown();
            }
        });

        //then
        assertTrue(latch.await(1, TimeUnit.SECONDS));
        assertEquals(1337, result.get());
    }

    @Test(expected = PubNubException.class)
    public void sync_whenFirstFails_RestIsNotCalled() throws PubNubException {
        firstDo(TestRemoteAction.failing())
                .then(integerResult -> {
                            fail("fail");
                            return TestRemoteAction.successful(15);
                        }
                ).sync();
    }

    @Test
    public void async_whenFirstFails_RestIsNotCalled() throws InterruptedException {
        //given
        final CountDownLatch latch = new CountDownLatch(1);
        final TestRemoteAction<Integer> successful = TestRemoteAction.successful(15);

        firstDo(TestRemoteAction.failing())
                .then(integerResult -> successful)
                //when
                .async((result, status) -> {
                    latch.countDown();
                });

        //then
        assertTrue(latch.await(1, TimeUnit.SECONDS));
        assertThat(successful.howManyTimesAsyncCalled(), is(0));
    }

    @Test
    public void cancel_cancelsCurrentlyRunningTask_RestIsNotCalled() throws InterruptedException {
        //given
        final CountDownLatch cancelSynchronisingLatch = new CountDownLatch(1);
        final CountDownLatch resultSynchronisingLatch = new CountDownLatch(1);
        final AtomicBoolean firstAsyncFinished = new AtomicBoolean(false);
        final CancellableRemoteAction<Object> longRunningTask = new CancellableRemoteAction<Object>() {
            @Override
            public void doAsync(@NotNull PNCallback<Object> callback) throws InterruptedException {
                cancelSynchronisingLatch.await();
                System.out.println("async");
                callback.onResponse(null, PNStatus.builder().build());
                firstAsyncFinished.set(true);
                resultSynchronisingLatch.countDown();
            }

            @Override
            public void silentCancel() {
                System.out.println("silentCancel");
                cancelSynchronisingLatch.countDown();
            }
        };
        final TestRemoteAction<Integer> successful = TestRemoteAction.successful(15);
        RemoteAction<Integer> composedAction = firstDo(longRunningTask).then(integerResult -> successful);
        composedAction.async((r, s) -> {
        });

        //when
        composedAction.silentCancel();

        //then
        assertTrue(resultSynchronisingLatch.await(1, TimeUnit.SECONDS));
        assertTrue(firstAsyncFinished.get());
        assertThat(successful.howManyTimesAsyncCalled(), is(0));
    }

    @Test
    public void retry_withoutCheckpointStartsFromBeginning() throws InterruptedException {
        //given
        CountDownLatch countDownLatch = new CountDownLatch(2);
        TestRemoteAction<Integer> firstSuccessful = TestRemoteAction.successful(1);
        TestRemoteAction<Integer> secondSuccessful = TestRemoteAction.successful(1);
        TestRemoteAction<Integer> firstFailing = TestRemoteAction.failingFirstCall(1);

        RemoteAction<Integer> composedAction = firstDo(firstSuccessful)
                .then(integerResult -> secondSuccessful)
                .then(integerResult -> firstFailing);

        //when
        composedAction.async((r, s) -> {
                    countDownLatch.countDown();
                    if (s.isError()) {
                        s.retry();
                    }
                }
        );

        //then
        assertTrue(countDownLatch.await(1000, TimeUnit.MILLISECONDS));
        assertThat(firstSuccessful.howManyTimesAsyncCalled(), is(2));
        assertThat(secondSuccessful.howManyTimesAsyncCalled(), is(2));
        assertThat(firstFailing.howManyTimesAsyncCalled(), is(2));
    }

    @Test
    public void retry_startsFromCheckpoint() throws InterruptedException {
        //given
        CountDownLatch countDownLatch = new CountDownLatch(2);
        TestRemoteAction<Integer> firstSuccessful = TestRemoteAction.successful(1);
        TestRemoteAction<Integer> secondSuccessful = TestRemoteAction.successful(1);
        TestRemoteAction<Integer> firstFailing = TestRemoteAction.failingFirstCall(1);

        RemoteAction<Integer> composedAction = firstDo(firstSuccessful)
                .checkpoint()
                .then(integerResult -> secondSuccessful)
                .then(integerResult -> firstFailing);

        //when
        composedAction.async((r, s) -> {
                    countDownLatch.countDown();
                    if (s.isError()) {
                        s.retry();
                    }
                }
        );

        //then
        assertTrue(countDownLatch.await(1000, TimeUnit.MILLISECONDS));
        assertThat(firstSuccessful.howManyTimesAsyncCalled(), is(1));
        assertThat(secondSuccessful.howManyTimesAsyncCalled(), is(2));
        assertThat(firstFailing.howManyTimesAsyncCalled(), is(2));
    }
}

File: src/test/kotlin/com/pubnub/api/legacy/endpoints/remoteaction/ComposableRemoteActionTest.kt
package com.pubnub.api.legacy.endpoints.remoteaction

import com.pubnub.api.PubNubException
import com.pubnub.api.endpoints.remoteaction.ComposableRemoteAction.Companion.firstDo
import com.pubnub.api.endpoints.remoteaction.RemoteAction
import com.pubnub.api.enums.PNOperationType
import com.pubnub.api.enums.PNStatusCategory
import com.pubnub.api.models.consumer.PNStatus
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.junit.Assert
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

class ComposableRemoteActionTest {
    @Test
    @Throws(PubNubException::class)
    fun sync_happyPath() {
        // given
        val composedAction: RemoteAction<Int> = firstDo(TestRemoteAction.successful(668))
            .then { TestRemoteAction.successful(it * 2) }
            .then { TestRemoteAction.successful(it + 1) }

        // when
        val result = composedAction.sync()!!

        // then
        Assert.assertEquals(1337, result.toLong())
    }

    @Test
    @Throws(InterruptedException::class)
    fun async_happyPath() {
        // given
        val latch = CountDownLatch(1)
        val result = AtomicInteger(0)
        val composedAction: RemoteAction<Int> = firstDo(TestRemoteAction.successful(668))
            .then { TestRemoteAction.successful(it * 2) }
            .then { TestRemoteAction.successful(it + 1) }

        // when
        composedAction.async { r: Int?, _: PNStatus? ->
            if (r != null) {
                result.set(r)
                latch.countDown()
            }
        }

        // then
        Assert.assertTrue(latch.await(1, TimeUnit.SECONDS))
        Assert.assertEquals(1337, result.get().toLong())
    }

    @Test(expected = PubNubException::class)
    @Throws(PubNubException::class)
    fun sync_whenFirstFails_RestIsNotCalled() {
        firstDo(TestRemoteAction.failing<Int>())
            .then {
                Assert.fail("fail")
                TestRemoteAction.successful(15)
            }.sync()
    }

    @Test
    @Throws(InterruptedException::class)
    fun async_whenFirstFails_RestIsNotCalled() {
        // given
        val latch = CountDownLatch(1)
        val successful: TestRemoteAction<Int> = TestRemoteAction.successful(15)
        firstDo(TestRemoteAction.failing<Int>())
            .then { successful } // when
            .async { _, _ -> latch.countDown() }

        // then
        Assert.assertTrue(latch.await(1, TimeUnit.SECONDS))
        assertThat(successful.howManyTimesAsyncCalled(), Matchers.`is`(0))
    }

    @Test
    @Throws(InterruptedException::class)
    fun cancel_cancelsCurrentlyRunningTask_RestIsNotCalled() {
        // given
        val cancelSynchronisingLatch = CountDownLatch(1)
        val resultSynchronisingLatch = CountDownLatch(1)
        val firstAsyncFinished = AtomicBoolean(false)
        val longRunningTask: CancellableRemoteAction<Any?> = object : CancellableRemoteAction<Any?> {
            @Throws(InterruptedException::class)
            override fun doAsync(callback: (result: Any?, status: PNStatus) -> Unit) {
                cancelSynchronisingLatch.await()
                println("async")
                callback(null, PNStatus(category = PNStatusCategory.PNAcknowledgmentCategory, error = false, operation = operationType()))
                firstAsyncFinished.set(true)
                resultSynchronisingLatch.countDown()
            }

            override fun silentCancel() {
                println("silentCancel")
                cancelSynchronisingLatch.countDown()
            }

            override fun operationType(): PNOperationType {
                return PNOperationType.FileOperation
            }
        }
        val successful: TestRemoteAction<Int> = TestRemoteAction.successful(15)
        val composedAction: RemoteAction<Int> = firstDo(longRunningTask).then { successful }
        composedAction.async { _: Int?, _: PNStatus? -> }

        // when
        composedAction.silentCancel()

        // then
        Assert.assertTrue(resultSynchronisingLatch.await(1, TimeUnit.SECONDS))
        Assert.assertTrue(firstAsyncFinished.get())
        assertThat(successful.howManyTimesAsyncCalled(), Matchers.`is`(0))
    }

    @Test
    @Throws(InterruptedException::class)
    fun retry_withoutCheckpointStartsFromBeginning() {
        // given
        val countDownLatch = CountDownLatch(2)
        val firstSuccessful: TestRemoteAction<Int> = TestRemoteAction.successful(1)
        val secondSuccessful: TestRemoteAction<Int> = TestRemoteAction.successful(1)
        val firstFailing: TestRemoteAction<Int> = TestRemoteAction.failingFirstCall(1)
        val composedAction: RemoteAction<Int> = firstDo(firstSuccessful)
            .then { secondSuccessful }
            .then { firstFailing }

        // when
        composedAction.async { _: Int?, s: PNStatus ->
            countDownLatch.countDown()
            if (s.error) {
                s.retry()
            }
        }

        // then
        Assert.assertTrue(countDownLatch.await(1000, TimeUnit.MILLISECONDS))
        assertThat(firstSuccessful.howManyTimesAsyncCalled(), Matchers.`is`(2))
        assertThat(secondSuccessful.howManyTimesAsyncCalled(), Matchers.`is`(2))
        assertThat(firstFailing.howManyTimesAsyncCalled(), Matchers.`is`(2))
    }

    @Test
    @Throws(InterruptedException::class)
    fun retry_startsFromCheckpoint() {
        // given
        val countDownLatch = CountDownLatch(2)
        val firstSuccessful: TestRemoteAction<Int> = TestRemoteAction.successful(1)
        val secondSuccessful: TestRemoteAction<Int> = TestRemoteAction.successful(1)
        val firstFailing: TestRemoteAction<Int> = TestRemoteAction.failingFirstCall(1)
        val composedAction: RemoteAction<Int> = firstDo(firstSuccessful)
            .checkpoint()
            .then { secondSuccessful }
            .then { firstFailing }

        // when
        composedAction.async { _: Int?, s: PNStatus ->
            countDownLatch.countDown()
            if (s.error) {
                s.retry()
            }
        }

        // then
        Assert.assertTrue(countDownLatch.await(1000, TimeUnit.MILLISECONDS))
        assertThat(firstSuccessful.howManyTimesAsyncCalled(), Matchers.`is`(1))
        assertThat(secondSuccessful.howManyTimesAsyncCalled(), Matchers.`is`(2))
        assertThat(firstFailing.howManyTimesAsyncCalled(), Matchers.`is`(2))
    }
}
